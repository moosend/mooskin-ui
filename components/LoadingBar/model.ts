import { IBoxComponentProps } from '../Box/model';

export interface ILoadingBarComponentProps extends IBoxComponentProps {
	/** progress of the loadingBar */
	progress: number;

	/** wether an error occured */
	error?: boolean;

	/** callback function to be fired when progress is done */
	onLoaderDone?: () => void;

	/** callback to be fired on error */
	onLoaderError?: () => void;
}
