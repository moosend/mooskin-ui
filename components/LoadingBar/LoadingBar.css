.loadingBar{
    transition: all 0.4s ease;
    position: fixed;
    top: 0;
    height: 3px;
    background: #F48770;
    box-shadow: -3px 0 15px 1px rgba(244,135,112, 0.4);
}

.error{
    background: #cc0000;
    box-shadow: -3px 0 15px 1px rgba(204,0,0, 0.4);
}

.loadingRight{
    left: 0;
}

.loadingLeft{
    right: 0;
}

.full{
    transition: all 0.1s ease;
    height: 0;
    opacity: 0;
}