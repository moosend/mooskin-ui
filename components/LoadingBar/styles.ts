import styled from 'styled-components';

// Models
import { ILoadingBarComponentProps } from './model';

// Components
import { Box } from '../Box/Box';

// "CSS" variables
import variables from '../_utils/globals/variables';
import { IStyledTheme } from '../index';

export const StyledLoadingBar = styled(Box)<ILoadingBarComponentProps>`
	position: fixed;
	top: 0;
	left: 0;
	height: 2px;
	width: ${(props) => `${props.progress}%`};
	box-shadow: ${(props) => `${getBgColor(props.error, props.palette)} 0px 0px 10px, ${getBgColor(props.error, props.palette)} 0px 0px 5px`};
	background-color: ${(props) => getBgColor(props.error, props.palette)};
	transition: all 500ms ease 0s;
`;

StyledLoadingBar.displayName = 'StyledLoadingBar';

const getBgColor = (error?: boolean, palette?: IStyledTheme) =>
	error
		? palette
			? palette.backgroundColors.primary4
			: variables.backgroundColors.primary4
		: palette
		? palette.backgroundColors.primary1
		: variables.backgroundColors.primary1;
