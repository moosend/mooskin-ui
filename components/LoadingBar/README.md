# MooSkin UI - LoadingBar Component

___

### Usage

To start using the LoadingBar Component first you have to Import it

```
Import {LoadingBar} from 'mooskin';
```
or modular import
```
// JS
import LoadingBar from 'mooskin/lib/LoadingBar';

// CSS
import 'mooskin/lib/LoadingBar/style.css';
```

And then you can simply start using it by typing

```
<LoadingBar attribute1="atr" attribute2="asd" />
```

Like most React components, it will accept given attributes and render differently based on the given attributes

### Examples

LoadingBar component requires a `progress` prop passed to it ranging from 0 to 100. Preferably passed from an application state.

```
<LoadingBar progress={this.state.progress} />
```

or pass it callback functions when the loader is completed, or when an error occurs

```
<LoadingBar progress={this.state.progress} error={this.state.someError} onProgressDone={yourFunc} onError={yourFunc} />
```

or just give it a custom style and class

```
<LoadingBar progress={this.state.progress} className="myClass" style={{color: 'green'}} />
```

<div class="playground-doc">

## Supported attributes

* `id` - id of the element
* `progress` - progress of the loading bar, numerical value from 0 to 100
* `direction` - direction of the loading bar (left or right)
* `className` - css class
* `style` - loading bar style
* `error` - boolean value, wether there's an error or not
* `onProgressDone` - callback to be triggered when loader is done (progress reached 100)
* `onError` - callback to be triggered when an error occurs (error prop is true)

</div>

Allthough these attributes are supported, only `progress` is mandatory.


#### For more

___

[MooSkin-UI](https://github.com/moosend/mooskin-ui)