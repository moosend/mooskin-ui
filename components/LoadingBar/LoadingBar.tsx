import * as React from 'react';

// Mooskin Context HoC that passes context to component props
import { withMooskinContext } from '../Styled/MooskinContextProvider';

// Models
import { ILoadingBarComponentProps } from './model';

// Styled Components
import { StyledLoadingBar } from './styles';

export const LoadingBar: React.FC<ILoadingBarComponentProps> = withMooskinContext((props) => {
	const [progress, setProgress] = React.useState(props.progress);
	const [opacity, setOpacity] = React.useState(1);

	React.useEffect(() => {
		setProgress(props.progress);
	}, [props.progress]);

	React.useEffect(() => {
		if (props.error) {
			props.onLoaderError && props.onLoaderError();
			setProgress(100);
		}
	}, [props.error]);

	React.useEffect(() => {
		if (progress >= 100) {
			!props.error && props.onLoaderDone && props.onLoaderDone();
			setTimeout(() => {
				setOpacity(0);
			}, 500);
			setTimeout(() => {
				setProgress(0);
			}, 1000);
			setTimeout(() => {
				setOpacity(1);
			}, 1500);
		}
	}, [progress]);

	return <StyledLoadingBar {...props} progress={progress} opacity={opacity} />;
});

LoadingBar.defaultProps = {
	className: '',
	progress: 0,
	style: {},
};

LoadingBar.displayName = 'LoadingBar';
