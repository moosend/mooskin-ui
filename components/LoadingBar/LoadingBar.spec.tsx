import * as React from 'react';

import { LoadingBar } from './LoadingBar';

import { mount } from 'enzyme';

describe('LoadingBar', () => {
	test(`LoadingBar renders correctly`, () => {
		const func1 = jest.fn();
		const func2 = jest.fn();

		const tree = mount(<LoadingBar progress={10} error={false} onLoaderError={func1} onLoaderDone={func2} />);

		expect(tree).toMatchSnapshot();
	});

	test(`calls done callback function on complete`, () => {
		const func = jest.fn();

		const component = mount(<LoadingBar progress={80} error={false} onLoaderDone={func} />);

		component.setProps({ progress: 100 });

		setTimeout(() => {
			expect(func).toHaveBeenCalled();
		}, 10);
	});

	test(`onLoaderError callback is called when an error prop of true is passed`, () => {
		const func = jest.fn();

		const component = mount(<LoadingBar progress={10} onLoaderError={func} />);

		component.setProps({ error: true });

		setTimeout(() => {
			expect(func).toHaveBeenCalled();
		}, 10);
	});
});
