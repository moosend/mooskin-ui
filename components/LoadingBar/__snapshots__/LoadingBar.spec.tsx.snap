// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingBar LoadingBar renders correctly 1`] = `
<div
  className="loading-component loadingBar loadingLeft   myClass"
  id="5"
  style={
    Object {
      "color": "red",
      "width": "10%",
    }
  }
/>
`;
